source/browseMode.py

```diff
@@ -2001,6 +2041,129 @@ def _iterNotLinkBlock(self, direction="next", pos=None):
 				yield TextInfoQuickNavItem("notLinkBlock", self, textRange)
 			item1=item2
 
+	STYLE_ATTRIBUTES = frozenset([
+		'background-color',
+		'color',
+		'font-family',
+		'font-size',
+		'bold',
+		'italic',
+		'marked',
+		'strikethrough',
+		'text-line-through-style',
+		'underline',
+		'text-underline-style',
+	])
+
+	def _extractStyles(
+			self,
+			info: textInfos.TextInfo,
+	) -> "textInfos.TextInfo.TextWithFieldsT":
+		"""
+		This function calls TextInfo.getTextWithFields(), and then processes fields in the following way:
+		1. Highlighted(marked) text is currently reported as Role.MARKED_CONTENT, and not formatChange.
+		For ease of further handling we create a new boolean format field "marked"
+		and set its value according to presence of Role.MARKED_CONTENT.
+		2. Then we drop all control fields, leaving only formatChange fields and text.
+		"""
+		stack = [{}]
+		result = []
+		for field in info.getTextWithFields():
+			if isinstance(field, textInfos.FieldCommand):
+				if field.command == 'controlStart':
+					style = {**stack[-1]}
+					if field.field.get('role') == controlTypes.Role.MARKED_CONTENT:
+						style['marked'] = True
+					stack.append(style)
+				elif field.command == 'controlEnd':
+					del stack[-1]
+				elif field.command == 'formatChange':
+					field.field = {
+						k: v
+						for k, v in {**field.field, **stack[-1]}.items()
+						if k in self.STYLE_ATTRIBUTES
+					}
+					result.append(field)
+				else:
+					raise RuntimeError()
+			elif isinstance(field, str):
+				result.append(field)
+			else:
+				raise RuntimeError
+		return result
+
+	def _iterTextStyle(
+			self,
+			kind: str,
+			direction: str = "next",
+			pos: textInfos.TextInfo = None
+	) -> Generator[TextInfoQuickNavItem, None, None]:
+		sameStyle = kind == 'sameStyle'
+		initialTextInfo = pos.copy()
+		initialTextInfo.collapse()
+		result = initialTextInfo.move(textInfos.UNIT_CHARACTER, 1, endPoint='end')
+		if result == 0:
+			result = initialTextInfo.move(textInfos.UNIT_CHARACTER, -1, endPoint='start')
+			if result == 0:
+				# Translators: Error message for same/different style quick navigation command
+				ui.message(_("Cannot determine current style"))
+				raise RuntimeError
+		styles = self._extractStyles(initialTextInfo)
+		if (
+			len(styles) == 0
+			or not isinstance(styles[0], textInfos.FieldCommand)
+			or styles[0].command != 'formatChange'
+		):
+			# Translators: Error message for same/different style quick navigation commands
+			ui.message(_("Cannot determine current style"))
+			raise RuntimeError
+		initialStyle = styles[0]
+
+		firstParagraph = True
+		paragraph = pos.copy()
+		tmpInfo = pos.copy()
+		tmpInfo.expand(textInfos.UNIT_PARAGRAPH)
+		paragraph.setEndPoint(tmpInfo, which='endToEnd' if direction == 'next' else 'startToStart')
+		while True:

```

Comment:
I'm concerned about being caught in `while True` 

```suggestion
		MAX_ITER_LIMIT = 10**6
		for _ in range(MAX_ITER_LIMIT):
```